### PR TITLE
Improvements to Typed Instances

### PR DESCRIPTION
**Pull request by erlendvollset**
TODO: Make CDM generator generate this

### CODE REVIEW

File: cognite/client/data_classes/data_modeling/instances.py [hunk 1]
```diff
@@ -1446,3 +1448,305 @@ def __delitem__(self, key: str | tuple[str, str] | tuple[str, str, str]) -> None
...
+        raise NotImplementedError
+
+    def _dump_properties(self) -> dict[str, Any]:
+        props = {key: prop for key, prop in vars(self).items() if key not in _RESERVED_PROPERTY_NAMES}
+        return _PropertyValueSerializer.serialize_values(props, camel_case=True)
+
+    @classmethod
+    def _load_instance(cls, resource: dict[str, Any], properties: dict[str, Any]) -> Self:
+        kwargs: dict[str, Any] = {}
+        signature = inspect.signature(cls.__init__)
```
Comment (doctrino) [inline cognite/client/data_classes/data_modeling/instances.py:1508]: Should be cached as it is an expensive call in a hot loop.

File: cognite/client/data_classes/data_modeling/instances.py [hunk 2]
```diff
@@ -1446,3 +1448,305 @@ def __delitem__(self, key: str | tuple[str, str] | tuple[str, str, str]) -> None
...
+        kwargs.update(cls._load_base_properties(resource, signature))
+        return cls(**kwargs)
+
+    @classmethod
+    def _load_base_properties(cls, resource: dict[str, Any], signature: inspect.Signature) -> dict[str, Any]:
+        args: dict[str, Any] = {}
+        for key in cls._base_properties():
+            camel_key = to_camel_case(key)
+            if camel_key in resource:
+                if key in signature.parameters and key in signature.parameters:
```
Comment (doctrino) [inline cognite/client/data_classes/data_modeling/instances.py:1519]:
```suggestion
                if key in signature.parameters:
```

File: cognite/client/data_classes/data_modeling/instances.py [hunk 3]
```diff
@@ -1446,3 +1448,305 @@ def __delitem__(self, key: str | tuple[str, str] | tuple[str, str, str]) -> None
...
+        external_id: str,
+        existing_version: int | None = None,
+        type: DirectRelationReference | tuple[str, str] | None = None,
+    ) -> None:
+        super().__init__(space, external_id, existing_version, None, type)
+
+    @staticmethod
+    @lru_cache(1)
+    def _base_properties() -> set[str]:
+        return _get_constructor_params(TypedEdge)
```
Comment (doctrino) [inline cognite/client/data_classes/data_modeling/instances.py:1604]:
```suggestion
        return _get_constructor_params(TypedNodeApply)
```

File: cognite/client/data_classes/data_modeling/instances.py [hunk 4]
```diff
@@ -1446,3 +1448,305 @@ def __delitem__(self, key: str | tuple[str, str] | tuple[str, str, str]) -> None
...
+        start_node: DirectRelationReference | tuple[str, str],
+        end_node: DirectRelationReference | tuple[str, str],
+        existing_version: int | None = None,
+    ) -> None:
+        super().__init__(space, external_id, type, start_node, end_node, existing_version, None)
+
+    @staticmethod
+    @lru_cache(1)
+    def _base_properties() -> set[str]:
+        return _get_constructor_params(TypedEdge)
```
Comment (doctrino) [inline cognite/client/data_classes/data_modeling/instances.py:1632]:
```suggestion
        return _get_constructor_params(TypedEdgeApply)
```

File: cognite/client/data_classes/data_modeling/instances.py [hunk 5]
```diff
@@ -1446,3 +1448,305 @@ def __delitem__(self, key: str | tuple[str, str] | tuple[str, str, str]) -> None
...
+        created_time: int,
+        deleted_time: int | None,
+        type: DirectRelationReference | tuple[str, str] | None,
+    ) -> None:
+        super().__init__(space, external_id, version, last_updated_time, created_time, deleted_time, None, type)
+
+    @staticmethod
+    @lru_cache(1)
+    def _base_properties() -> set[str]:
+        return _get_constructor_params(TypedEdge)
```
Comment (doctrino) [inline cognite/client/data_classes/data_modeling/instances.py:1661]:
```suggestion
        return _get_constructor_params(TypedNode)
```

File: cognite/client/data_classes/data_modeling/cdm/v1.py [hunk 6]
```diff
@@ -406,14 +406,14 @@ def __init__(
...
-        TypedNodeApply.__init__(self, space, external_id, existing_version, None, type)
+        TypedNodeApply.__init__(self, space, external_id, existing_version, type)
         self.name = name
         self.description = description
         self.tags = tags
         self.aliases = aliases
 
 
-class CogniteDescribableNode(_CogniteDescribableProperties, TypedNode):
+class CogniteDescribable(_CogniteDescribableProperties, TypedNode):
```
Comment (doctrino) [inline cognite/client/data_classes/data_modeling/cdm/v1.py:416]: Is the thinking that the node version of Describable is without suffix node, while we keep it for edge?

**Review by doctrino — COMMENTED** [review]
Nice cleanup. One question and what looks like a few copy-paste issues.

File: cognite/client/data_classes/data_modeling/cdm/v1.py (hunk 6 above)
Comment (erlendvollset) [reply cognite/client/data_classes/data_modeling/cdm/v1.py:416]: Ohh. I didn't realize we had a separate one for Edge. My bad, then it makes sense to keep it for both. I'll revert this commit.

File: cognite/client/data_classes/data_modeling/instances.py (hunk 1 above)
Comment (erlendvollset) [reply cognite/client/data_classes/data_modeling/instances.py:1508]: Good point. Will fix.

File: cognite/client/data_classes/data_modeling/instances.py (hunk 2 above)
Comment (haakonvt) [reply cognite/client/data_classes/data_modeling/instances.py]: Twice as safe to cosmic rays got to be worth something 😂

File: cognite/client/data_classes/data_modeling/instances.py [hunk 7]
```diff
@@ -1446,3 +1448,305 @@ def __delitem__(self, key: str | tuple[str, str] | tuple[str, str, str]) -> None
...
+            start_node,
+            end_node,
+            deleted_time,
+            properties=None,
+        )
+
+    @staticmethod
+    @lru_cache(1)
+    def _base_properties() -> set[str]:
+        return _get_constructor_params(TypedEdge)
```
Comment (haakonvt) [inline cognite/client/data_classes/data_modeling/instances.py]: You already cache `_get_constructor_params` btw

**Review by doctrino — APPROVED** [review]

File: cognite/client/data_classes/data_modeling/instances.py (hunk 7 above)
Comment (erlendvollset) [reply cognite/client/data_classes/data_modeling/instances.py]: Addressed 👍

**Pull request merged**